internal/metrics/nonblocking_metrics.go

```diff
@@ -0,0 +1,254 @@
+func (nbmw *NonblockingMetricsWrapper) NewMetricVec(name string, help string, labelNames ...string) (commontypes.MetricVec, error) {
+	if err := validName(name); err != nil {
+		return nil, fmt.Errorf("could not create metric vector with invalid name: %w", err)
+	}
+
+	if err := validLabelNames(labelNames); err != nil {
+		return nil, fmt.Errorf("could not create metric vector with invalid label names: %w", err)
+	}
+
+	return &nonblockingMetricVec{
+		nbmw,
+		name,
+		help,
+		labelNames,
+		nil, // initialized in flush loop
+	}, nil
+}
+
+func validLabelNames(labelNames []string) error {
+	for _, name := range labelNames {
+		if err := validName(name); err != nil {
+			return fmt.Errorf("format of label name %q is not valid: %w", name, err)
+		}
+	}
+	return nil
+}
+
+// Close shuts down the flush loop and waits for the go routines of the NonblockingMetricsWrapper to finish
+// Close does not close the underlying metrics implementation. It should be the responsibility of the underlying
+// metrics implementation to shut down without go routines leakage
+func (nbmw *NonblockingMetricsWrapper) Close() error {
+	nbmw.logger.Debug("NonblockingMetricsWrapper: Closing non-blocking metrics wrapper", nil)
+	nbmw.ctxCancel()
+	nbmw.subs.Wait()
+	nbmw.logger.Info("NonblockingMetricsWrapper: Closed non-blocking metrics wrapper", nil)
+	return nil
+}
+
+func (nbmw *NonblockingMetricsWrapper) flushLoop() {
+	nbmw.logger.Debug("NonblockingMetricsWrapper: Starting metric reporter flush loop", nil)
+	ticker := time.NewTicker(nbmw.flushInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-nbmw.ctx.Done():
+			nbmw.logger.Debug("NonblockingMetricsWrapper: Canceling metric reporter flush loop", nil)
+			return
+		case <-ticker.C:
+			nbmw.flush()
+		}
+	}
+}
+
+func (nbmw *NonblockingMetricsWrapper) flush() {
+	nbmw.logger.Trace("NonblockingMetricsWrapper: Flushing metricsImpl", nil)
+	nbmw.metricIndex.Range(
+		func(fingerprint string, nonblockingMetric *nonblockingMetric) bool {
+			// The underlying metric implementation might not have been initialized yet
+			if nonblockingMetric.metricImpl == nil {
+				// Get the parent metric vector
+				nonblockingMetricVec := nonblockingMetric.nonblockingMetricVec
+				// The underlying metric vector implementation might not have been initialized yet
+				if nonblockingMetricVec.metricVecImpl == nil {
+					var err error
+					nonblockingMetricVec.metricVecImpl, err = nbmw.metricsImpl.NewMetricVec(nonblockingMetricVec.name, nonblockingMetricVec.help, nonblockingMetricVec.labelNames...)
+					if err != nil {
+						// TODO: This will potentially log the same error in every flush loop iteration, consider deduplicating.
+						nbmw.logger.Error("NonblockingMetricsWrapper: Could not create metric vector", commontypes.LogFields{"err": err.Error()})
+						return true
+					}
+				}
+				var err error
+				nonblockingMetric.metricImpl, err = nonblockingMetricVec.metricVecImpl.GetMetricWith(nonblockingMetric.labels)
```

Comment:
This can segfault if nonblockingMetric is nil.  Should we add a nil check, so it's backward compatible with PeerConfig structs that don't have Metrics field explicitly filled in?

Reply:
Yeah, we should def. do that.  Note that this is just an alpha branch with the goal of aligning on the updated libocr interface and being able to do some very rough testing of whether things work.

Reply:
Do you refer to line 122?
If yes, do I understand correctly that the nil check you refer to is for the constructor?
I.e. function "NewNonblockingMetricsWrapper"

Reply:
Yes, line 122.  I just mean adding:
```
if nbmw.metricsImpl == nil {

}
```
IIRC, when I left this comment I did a couple edits within the first minute or two... you might have seen an earlier version in an email that wasn't right.

Reply:
Oh, I may have misunderstood the question:  I was actually thinking the check should go here, but yes putting in the constructor might work better?  I'd need more context to have a strong opinion on that.